test/functional/lua/vim_spec.lua

```diff
@@ -128,33 +128,93 @@ describe('lua stdlib', function()
   end)
 
-  it('vim.deprecate', function()
+  local function test_vim_deprecate(current_version)
     -- vim.deprecate(name, alternative, version, plugin, backtrace)
-    eq(
-      dedent [[
-      foo.bar() is deprecated, use zub.wooo{ok=yay} instead. :help deprecated
-      This feature will be removed in Nvim version 0.10]],
-      exec_lua('return vim.deprecate(...)', 'foo.bar()', 'zub.wooo{ok=yay}', '0.10')
-    )
-    -- Same message, skipped.
-    eq(vim.NIL, exec_lua('return vim.deprecate(...)', 'foo.bar()', 'zub.wooo{ok=yay}', '0.10'))
-    -- Don't show error if not hard deprecated
-    eq(vim.NIL, exec_lua('return vim.deprecate(...)', 'foo.bar()', 'nil', '5000.0.0'))
-    -- When `plugin` is specified, don't show ":help deprecated". #22235
-    eq(
-      dedent [[
-      foo.bar() is deprecated, use zub.wooo{ok=yay} instead.
-      This feature will be removed in my-plugin.nvim version 0.3.0]],
-      exec_lua(
-        'return vim.deprecate(...)',
-        'foo.bar()',
-        'zub.wooo{ok=yay}',
-        '0.3.0',
-        'my-plugin.nvim',
-        false
-      )
-    )
-  end)
+    -- See MAINTAIN.md for the soft/hard deprecation policy
+
+    describe(('vim.deprecate [current_version = %s]'):format(current_version), function()
+      before_each(function()
+        -- mock vim.version() behavior, should be pinned for consistent testing
+        exec_lua(
+          [[
+            local current_version_mock = vim.version.parse(...)
+            getmetatable(vim.version).__call = function()
+              return current_version_mock
+            end
+          ]],
+          current_version
+        )
+      end)
+
+      it('when plugin = nil', function()
+        eq(
+          dedent [[
+            foo.bar() is deprecated, use zub.wooo{ok=yay} instead. :help deprecated
+            This feature will be removed in Nvim version 0.10]],
+          exec_lua('return vim.deprecate(...)', 'foo.bar()', 'zub.wooo{ok=yay}', '0.10')
+        )
+        -- Same message, skipped.
+        eq(vim.NIL, exec_lua('return vim.deprecate(...)', 'foo.bar()', 'zub.wooo{ok=yay}', '0.10'))
+
+        -- Don't show error if not hard-deprecated (only soft-deprecated)
+        eq(
+          vim.NIL,
+          exec_lua('return vim.deprecate(...)', 'foo.baz()', 'foo.better_baz()', '0.12.0')
+        )
+
+        -- Show error if hard-deprecated
+        eq(
+          dedent [[
+            foo.hard_dep() is deprecated, use vim.new_api() instead. :help deprecated
+            This feature will be removed in Nvim version 0.11]],
+          exec_lua('return vim.deprecate(...)', 'foo.hard_dep()', 'vim.new_api()', '0.11')
+        )
+
+        -- To be deleted in the next major version (1.0)
+        eq(
+          dedent [[
+            foo.baz() is deprecated. :help deprecated
+            This feature will be removed in Nvim version 1.0]],
+          exec_lua [[ return vim.deprecate('foo.baz()', nil, '1.0') ]]
+        )
```

Comment:
Wait, this is sus. Is it supposed to give a deprecation warning for 1.0?

Reply:
It's a corner case that was missed previously. Previously (current master), the behavior is to not show the warning (although we never had to-be-deprecated-in-1.0 yet): `if version < {1, -1}`, but actually warnings won't shown until we bump up the version to 1.0-dev. As I wrote in the comment, I assumed we don't call `vim.deprecate(..., '1.0')` until our development cycle is towards 1.0 (i.e. current version, say 0.20 or whatever, is the "last" 0.x version) to detect it by any mistake.

For example, if v0.20 is the last 0.x version, then a feature that is deprecated in 0.19-dev should have:
- soft-deprecation through v0.19, and
- hard-deprecation in v0.20 (show warnings!)
- removal in v1.0

but when the master is bumped up to v0.20, warning will be still suppressed. It's tricky because we can't programmatically know whether this is the last minor version planned. 

Actually we *do* have some plans to remove things in the next major version, so I'm fine with either behavior, I'm fine with reverting it back.